src/tangent_types/abstract_zero.jl

```diff
@@ -87,3 +87,32 @@ arguments.
 ```
 """
 struct NoTangent <: AbstractZero end
+
+"""
+    zero_tangent(primal)
+
+This returns an appropriate zero tangent suitable for accumulating tangents of the primal.
+For mutable composites types this is a structural []`MutableTangent`](@ref)
+For `Array`s, it is applied recursively for each element.
+For immutable types, this is simply [`ZeroTangent()`](@ref) as accumulation is default out-of-place for contexts where mutation does not apply.
+(Where mutation is not to be supported even for mutable types, then [`ZeroTangent()`](@ref) should be used for everything)
+
+!!! warning Exprimental
+    `zero_tangent`is an experimental feature, and is part of the mutation support featureset.
+    While this notice remains it may have changes in behavour, and interface in any _minor_ version of ChainRulesCore.
+    Exactly how it should be used (e.g. is it forward-mode only?)
+"""
+function zero_tangent end
+zero_tangent(::AbstractString) = ZeroTangent()
+# zero_tangent(::Number) = zero(x) # TODO: do we want this?
+zero_tangent(primal::Array{<:Number}) = zero(primal)  # TODO: do we want this?
+zero_tangent(primal::Array) = map(zero_tangent, primal)
```

Comment:
Can you go into more detail on both your points?

Reply:
Ofc.

Suppose that `primal isa Vector{Vector{Float64}}`. If not every element of `primal` has been given a value, say, because `primal` was created using `Vector{Vector{Float64}}(undef, 10)` and not every element has been written to, `map(zero_tangent, primal)` will yield the following flavour of error:
```julia
ERROR: UndefRefError: access to undefined reference
Stacktrace:
 [1] getindex(A::Vector{Vector{Float64}}, i1::Int64)
   @ Base ./essentials.jl:13
 [2] top-level scope
   @ REPL[2]:1
```
In order to avoid this, `isassigned(primal, n)` must be checked before calling `getindex(primal, n)`. So I am advocating for ensuring that this happens.

The exception to this is when `isbitstype(eltype(primal))`. In this case the value of the elements is non-deterministic (I guess it depends on whatever happened to by lying around in the memory allocated to `primal`), but it's perfectly legal to call `getindex(primal, n)` for any `n`, so `map(zero_tangent, primal)` ought to succeed.

I'm imagining the following function:
```julia
function zero_tangent(x::Array{P, N}) where {P, N}
    isbitstype(P) && return map(zero_tangent, x)
    y = Array{tangent_type(P), N}(undef, size(x)...)
    @inbounds for n in eachindex(y)
        if isbitstype(P) || isassigned(x, n)
            y[n] = f(x[n])
        end
    end
    return y
end
```
where you would have to figure out how to implement `tangent_type`, or find some other way of determining the type of `y`.

Alternatively, we could just throw an error if someone tries to call `zero_tangent` on an `Array` with `undef` elements 🤷

Reply:
What should we put in the place of the `undef`? I see in your example you leave them `undef` also
but  its unclear exactly what we will be able to do with this.
Since like `+` is not defined for vectors containing undef elements.
It throws an `UndefRefError`.
So maybe for now we should just test that we do the same.
Turning an error into a useful output later is nonbreaking

Reply:
Main use of `zero_tangent` is for creating the partial for nonactive local variables that will later be mutated to contain something active
So  guess if we created it with undef, then if something is being taken out of it, then that something must have been put in, so the entry would no longer be undef by the time it is needed.
So putting an undef in is indeed fine. 
less fine if it is actually being returned, but if you are returning it from your function that also means you are returning a vector with undefs in it from your function so your life is already not great